File: src/utils/automodCache.js
const config = require('../../config/config.js');

const ESCAPE_REGEX = /[.*+?^${}()|[\]\\]/g;

class BannedWordCache {
    constructor() {
        this.invalidate();
    }

    invalidate() {
        this.signature = null;
        this.regex = null;
        this.highlightRegex = null;
        this.words = [];
    }

    _build() {
        const words = (config.autoModeration?.bannedWords || [])
            .map(word => (typeof word === 'string' ? word.trim() : ''))
            .filter(Boolean);

        const signature = words.join('|');
        if (signature === this.signature) {
            return;
        }

        this.signature = signature;
        this.words = words;

        if (!words.length) {
            this.regex = null;
            this.highlightRegex = null;
            return;
        }

        const escaped = words
            .map(word => word.replace(ESCAPE_REGEX, '\\$&'))
            .join('|');

        this.regex = new RegExp(`(${escaped})`, 'gi');
        this.highlightRegex = new RegExp(`(${escaped})`, 'gi');
    }

    getMatches(content) {
        if (!content) return [];
        this._build();
        if (!this.regex) return [];
        const matches = content.match(this.regex);
        if (!matches) return [];
        return matches.map(word => word.toLowerCase());
    }

    highlight(content) {
        if (!content) return '*Sin contenido*';
        this._build();
        if (!this.highlightRegex) return content;
        return content.replace(this.highlightRegex, '**$1**');
    }
}

module.exports = new BannedWordCache();

File: pybot/utils/automod_cache.py
from __future__ import annotations

import re
from typing import List

from .. import config

ESCAPE_REGEX = re.compile(r"[.*+?^${}()|[\]\\]")


class BannedWordCache:
    def __init__(self) -> None:
        self.invalidate()

    def invalidate(self) -> None:
        self.signature: str | None = None
        self.regex: re.Pattern[str] | None = None
        self.highlight_regex: re.Pattern[str] | None = None
        self.words: List[str] = []

    def _build(self) -> None:
        cfg = config.resolve_config()
        words = [word.strip() for word in cfg.get("autoModeration", {}).get("bannedWords", []) if isinstance(word, str)]
        signature = "|".join(words)
        if signature == self.signature:
            return

        self.signature = signature
        self.words = words

        if not words:
            self.regex = None
            self.highlight_regex = None
            return

        escaped = "|".join(ESCAPE_REGEX.sub(r"\\\\$&", word) for word in words)
        self.regex = re.compile(f"({escaped})", re.IGNORECASE)
        self.highlight_regex = re.compile(f"({escaped})", re.IGNORECASE)

    def get_matches(self, content: str) -> List[str]:
        if not content:
            return []
        self._build()
        if not self.regex:
            return []
        return [match.lower() for match in self.regex.findall(content)]

    def highlight(self, content: str) -> str:
        if not content:
            return "*Sin contenido*"
        self._build()
        if not self.highlight_regex:
            return content
        return self.highlight_regex.sub(r"**\\1**", content)


automod_cache = BannedWordCache()
